zebROS_ws/src/behaviors/src/game_piece_path_gen.cpp

```diff
@@ -0,0 +1,137 @@
+#include "field_obj/Detection.h"
+#include "trajectory_msgs/JointTrajectoryPoint.h"
+
+// Create service clients
+ros::ServiceClient spline_gen_cli;
+ros::ServiceClient dynamic_path_cli;
+
+field_obj::Detection lastObjectDetection;
+
+// Callback function to retrieve the most recent object detection message
+void objectDetectCallback(field_obj::DetectionConstPtr msg)
+{
+	lastObjectDetection = *msg;
+}
+
+trajectory_msgs::JointTrajectoryPoint generateTrajectoryPoint(double x, double y, double rotation) {
+	trajectory_msgs::JointTrajectoryPoint point;
+	point.positions.resize(3);
+	point.positions[2] = rotation;
+	point.positions[0] = x;
+	point.positions[1] = y;
+	return point;
+}
+
+bool genPath(behavior_actions::GamePiecePickup::Request& req, behavior_actions::GamePiecePickup::Response& res)
+{
+	res.success = false; // Default to failure
+  base_trajectory_msgs::GenerateSpline spline_gen_srv;
+
+  spline_gen_srv.request.header = lastObjectDetection.header;
+
+  const size_t objects_num = std::min((int)(lastObjectDetection.objects.size()), (int)(req.max_objects));
+	ROS_INFO_STREAM("game_piece_path_gen : objects_num: " << objects_num);
+
+  std::vector<std::array<double, 3>> points;
+  points.push_back({0, 0, 0}); // may need to worry about transforms between beginning/end point and intake?
+
+	for (size_t i = 0; i < objects_num; i++) // filter out selected object detections
+		if(lastObjectDetection.objects[i].id == req.object_id)
+			points.push_back({lastObjectDetection.objects[i].location.x, lastObjectDetection.objects[i].location.y, 0});
+
+	if (points.size() == 1) // No objects added
+	{
+		ROS_INFO_STREAM("game_piece_path_gen : no " << req.object_id << " objects detected");
+		res.message = "no power cells detected";
+		return false;
+	}
+
+  std::sort(points.begin(), points.end()); // sort on x coordinate
+
+  points.push_back({req.endpoint.position.x, req.endpoint.position.y, req.endpoint.orientation.z});
+
+  size_t points_num = points.size();
+  spline_gen_srv.request.points.resize(3*points_num-4); // 3 * (point_num - 2) + 2
+  spline_gen_srv.request.point_frame_id.resize(3*points_num-4);
+	size_t point_index = 0;
+	for (size_t i = 0; i < points_num; i++) // copy points into spline request
+	{
+		ROS_INFO_STREAM("(" << points[i][0] << ", " << points[i][1] << "), orientation: " << points[i][2]);
+		if ((i == 0) || (i == (points_num - 1)))
+		{
+			// Facing forward for the last point increases our chances
+			// of running over the ball
+			spline_gen_srv.request.points[point_index] = generateTrajectoryPoint(points[i][0], points[i][1], points[i][2]);
+			spline_gen_srv.request.point_frame_id[point_index] = "intake";
+			base_trajectory_msgs::PathOffsetLimit path_offset_limit;
+			spline_gen_srv.request.path_offset_limit.push_back(path_offset_limit);
+			point_index++;
+		}
+		else
+		{
+			spline_gen_srv.request.points[point_index] = generateTrajectoryPoint(points[i][0] - .175, points[i][1], points[i][2]);
```

Comment:
I'd also like to see a similar global var + config value for this .175 value
One of the TODOs is to tweak the code to make it more reliably pick up objects, and this seems like an obvious parameter which we might change
For extra credit points, make it dynamically reconfigurable.  The code to do that is in the ddynamic_reconfigure package.  You'd have to add this to the CMakeLists.txt catkin package list, then `#include <ddynamic_reconfigure/ddynamic_reconfigure.h>`, then follow the first example in the docs here : https://github.com/FRC900/ddynamic_reconfigure.
Basically, create a ddynamic_reconfigure::DDynamicReconfigure object, register a double var with it using ddr.registerVariable<int>, then call  ddr.publishServicesTopics().
Would make it easy to change the value on the fly and see what happens to the paths.

Reply:
Maybe we could add an array with a length that people can specify to say how many points to add for each game piece and what their offsets should be?

Reply:
Yeah, I'm not sure the exact correct way to go right now. We have a task which is along the lines of "optimize path used to pick up game pieces" - right now the number of points to hit is kinda hacky, and very dependent on how good the intake is (i.e. if we can hit balls from an angle, how wide it is, and so on)

I like the idea of generalizing it into an array - that would fit the current approach and would give some flexability for the future. Worst case we find it isn't general enough, which is no big deal, we just make it better next time we touch the code.  I think it is worth switching to, though, so go ahead with that idea.

I'm not sure how that fits into make it dynamic reconfigure, though - a variable length array doesn't really fit into how dynamic reconfigure works.  So ignore that part for now, we can revisit it later.